### PR TITLE
OpenConfig introduction

### PR DESCRIPTION
**Pull request by FlorisCO**
Jeroen Neyt and myself collaborated to create the documentation about OpenConfig, the CommunicationGateway DxM, and the middleware. We also expect a review by Pedro Debevere and Thomas Gunkel.
Note that the page introducing OpenConfig is mostly a copy of Pedro's page at [https://community.dataminer.services/expert-hub-openconfig](https://community.dataminer.services/expert-hub-openconfig)

So we are expecting some minor edits and perhaps pages moving in the toc.yml.

### CODE REVIEW

**Comment by FlorisCO** [comment]
Note, build fails due to cross-references to the API documentation that is part of another branch and task. These will be resolved when that documentation has been merged.

File: assemblies/DataSources/OpenConfig/Skyline.DataMiner.Helper.OpenConfig.xml [hunk 1]
```diff
@@ -1060,8 +1700,8 @@
         </member>
         <member name="T:Gnmi.Notification">
```
Comment (mmphs) [inline assemblies/DataSources/OpenConfig/Skyline.DataMiner.Helper.OpenConfig.xml:1701]: Everything below here doesn't have to be reviewed as it's 3rd party generated documentation. (Gnmi + GnimExt namespace)
It's excluded in our doc generation so it's not being displayed.

File: develop/DSI/DSIOpenConfig/Introduction.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+This RPC can be used by clients to obtain more information from a target about its capabilities. The capabilities comprise:
+
+* The version of the gNMI protocol (e.g. 0.7.0)
+* The supported encodings (e.g. Json, Proto)
+* The supported data models: A list of all the supported models. Each entry defines the name of the model, the version of the model and the organization that published the model
+
+> [!NOTE]
+> The list of supported models is not restricted to OpenConfig models. For example, implementors typically also define their own proprietary YANG models.
+
+#### get
```
Comment (PedroDebevere) [inline develop/DSI/DSIOpenConfig/Introduction.md:170]:
```suggestion
#### Get
```

File: develop/DSI/DSIOpenConfig/Introduction.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+The Get RPC can be used by clients to obtain a snapshot of the requested data (the data requested is provided as a number of paths from the data model tree).
+
+Note that a Get RPC should only be used to obtain small amounts of data from the target. To request larger data sets, it is advised to use a subscription.
+
+#### set
```
Comment (PedroDebevere) [inline develop/DSI/DSIOpenConfig/Introduction.md:176]:
```suggestion
#### Set
```

File: develop/DSI/DSIOpenConfig/Introduction.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+The Set RPC can be used to modify the state of a target. A set request supports the deletion, replacement and updating of items. All changes that are defined in a set request are considered as one transaction.
+
+#### subscribe
```
Comment (PedroDebevere) [inline develop/DSI/DSIOpenConfig/Introduction.md:180]:
```suggestion
#### Subscribe
```

File: assemblies/DataSources/OpenConfig/Skyline.DataMiner.Helper.OpenConfig.xml [hunk 5]
```diff
@@ -466,24 +772,47 @@
...
             <param name="inventory">A mapping of the path in the YANG model to the data entity.</param>
             <exception cref="T:System.ArgumentNullException">Exception that will be raised when one of the passed arguments is empty.</exception>
         </member>
-        <member name="P:Skyline.DataMiner.Helper.OpenConfig.DataMapper.DataMinerConnectorDataMapper.Inventory">
+        <member name="M:Skyline.DataMiner.Helper.OpenConfig.DataMapper.DataMinerConnectorDataMapper.RegisterSubscriptionPaths(System.String,System.Collections.Generic.IEnumerable{Gnmi.Path})">
+            <summary>
+            Registers the parameter IDs that are linked with a subscription name.
+            </summary>
+            <param name="subscriptionName">Name of the subscription.</param>
+            <param name="paths">List with paths that are subscribed.</param>
```
Comment (MariekeGO) [inline assemblies/DataSources/OpenConfig/Skyline.DataMiner.Helper.OpenConfig.xml:788]:
```suggestion
            <param name="paths">List of paths that are subscribed.</param>
```

File: assemblies/DataSources/OpenConfig/Skyline.DataMiner.Helper.OpenConfig.xml [hunk 6]
```diff
@@ -502,65 +831,110 @@
...
+            Reads out the existing values and stores them internally.
             </summary>
-            <param name="notification">gNMI notification to be processed.</param>
+        </member>
+        <member name="M:Skyline.DataMiner.Helper.OpenConfig.DataMapper.IDataMapper.RegisterSubscriptionPaths(System.String,System.Collections.Generic.IEnumerable{Gnmi.Path})">
+            <summary>
+            Registers the <see cref="T:Skyline.DataMiner.Helper.OpenConfig.DataMapper.IDataEntity"/> objects that are linked with a subscription name.
+            </summary>
+            <param name="subscriptionName">Name of the subscription.</param>
+            <param name="paths">List with paths that are subscribed.</param>
```
Comment (MariekeGO) [inline assemblies/DataSources/OpenConfig/Skyline.DataMiner.Helper.OpenConfig.xml:937]:
```suggestion
            <param name="paths">List of paths that are subscribed.</param>
```

File: assemblies/DataSources/OpenConfig/Skyline.DataMiner.Helper.OpenConfig.xml [hunk 7]
```diff
@@ -574,6 +948,196 @@
...
+        </member>
+        <member name="M:Skyline.DataMiner.Helper.OpenConfig.DataMapper.IDataMinerConnectorDataGridColumn.AddReferencedColumn(Skyline.DataMiner.Helper.OpenConfig.DataMapper.IDataMinerConnectorDataGridColumn)">
+            <summary>
+            Adds the reference of the column that is linked with this column.
+            </summary>
+            <param name="referencedColumn">Column that is linked with this column.</param>
+        </member>
+        <member name="M:Skyline.DataMiner.Helper.OpenConfig.DataMapper.IDataMinerConnectorDataGridColumn.GetReferencedColumns">
+            <summary>
+            Gets the list with column parameter IDs that have a reference to this column.
```
Comment (MariekeGO) [inline assemblies/DataSources/OpenConfig/Skyline.DataMiner.Helper.OpenConfig.xml:1068]:
```suggestion
            Gets the list of column parameter IDs that have a reference to this column.
```

File: develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+
+### Including the middleware package
+
+Include the NuGet package *Skyline.DataMiner.DataSources.OpenConfig.Gnmi* in your project, as explained under [Consuming NuGet packages](xref:Consuming_NuGet).
+
+When you are developing a connector, there is an optional package when you want to use the [DataMapper](#datamapper): *Skyline.DataMiner.DataSources.OpenConfig.Gnmi.Protocol*
+
+### Setting up the connection with the endpoint
+
+#### Creating a GnmiClient
```
Comment (MariekeGO) [inline develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md:31]: Adding more structure to the text underneath this header would make it more user-friendly. I recommend dividing this into subsections, for example one for each item that has to be passed.

File: develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+         disposed = true;
+         gnmiClient?.Dispose();
+         gnmiClient = null;
+      }
+   }
+}
+```
+
+> [!NOTE]
+> Making a QAction implement IDisposable is only supported in DataMiner version 10.2.9 and up.
```
Comment (MariekeGO) [inline develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md:181]: Is this note really necessary, since we've already said that the minimum version for everything here is 10.3.3?

File: develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+IDataMapper dataMapper = new DataMinerConnectorDataMapper(
+    protocol,
+    new List<IDataEntity>
+    {
+        new DataMinerConnectorParameter("system/memory/state/physical", Parameter.systemmemorystatephysical_400),
+        new DataMinerConnectorParameter("system/memory/state/reserved", Parameter.systemmemorystatereserved_404)
+    });
+```
+
+Having this configured as soon as a gNMI notification will come in with data for either the *system/memory/state/physical* or the *system/memory/state/reserved* YANG path, it will automatically do the set on the configured parameters. You need to pass it the `SLProtocol` object so it can execute those sets.
```
Comment (MariekeGO) [inline develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md:315]: This sentence is not clear. Can you rephrase this? In the very least, I'd get rid of the "Having this configured" structure and replace it with something less ambiguous.

File: develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+        {
+            new DataMinerConnectorDataGridColumn("openconfig-interfaces:type", Parameter.Interfacesstate.Pid.interfacesstatetype),
+            new DataMinerConnectorDataGridColumn("openconfig-interfaces:mtu", Parameter.Interfacesstate.Pid.interfacesstatemtu),
+            new DataMinerConnectorDataGridColumn("openconfig-interfaces:loopback-mode", Parameter.Interfacesstate.Pid.interfacesstateloopbackmode)
+        }
+    });
+```
+
+> [!IMPORTANT]
+> When configuring the path for a column, always specify the YANG module name as well. Notifications of type `JSON` do not contain them but notifications of type `JSON_IETF` do. The DataMapper is capable of handling both, but for that reason the YANG module name needs to be known.
```
Comment (MariekeGO) [inline develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md:334]: What's "them" here? If it's the YANG module name, the plural doesn't really make sense given the singular "YANG module name" in the preceding sentence, so I'd suggest writing "it" instead.

**Review by MariekeGO — CHANGES_REQUESTED** [review]

File: user-guide/Reference/DataMiner_Extension_Modules.md [hunk 12]
```diff
@@ -48,18 +48,26 @@ The SupportAssistant module is responsible for features allowing Skyline Communi
...
+The CommunicationGateway module makes it possible for connectors or scripts running in the DataMiner environment to communicate with devices that require a gRPC connection.
 
-1. Next to *ModelHost*, select to deploy this DxM.
+At present, we support the [gNMI](xref:DSI_OpenConfig_Introduction#gnmi) service, but more services and RPC connections may be added.
 
-### UserDefinableApiEndpoint
+The DxM exists in tandem with service-specific [middleware](xref:Nuget_Communication_Middleware) that facilitates the NATS communication to the DxM. However, not all middleware require the CommunicationGateway to establish a connection with the device. The DxM primarily exists to bridge the gap between different dotNet versions. Namely, gRPC requires dotNet 5 or higher for full functionality, while SLScripting still runs on dotNet 4.6.2.
 
-The UserDefinableApiEndpoint module is responsible for handling API triggers for user-defined APIs. See [DataMiner UserDefinableApiEndpoint DxM](xref:UD_APIs_UserDefinableApiEndpoint).
+#### Prerequisites
```
Comment (MariekeGO) [inline user-guide/Reference/DataMiner_Extension_Modules.md:65]: The information from line 65 onwards is not in the right place. This page should only contain general information about the DxMs, and should link to other pages for more specific information on how to actually use them. I recommend linking to DSI_OpenConfig_Middleware and moving the information from "Prerequisites" and "Getting started" to that page if it isn't included yet there.

File: assemblies/DataSources/OpenConfig/Skyline.DataMiner.DataSources.OpenConfig.Gnmi.Protocol.xml
Comment (mmphs) [inline assemblies/DataSources/OpenConfig/Skyline.DataMiner.DataSources.OpenConfig.Gnmi.Protocol.xml:1]: This file is new but contains only content that has been reviewed yet. It has been extracted in a separate file because it's in a different library.

File: develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md (hunk 9 above)
Comment (mmphs) [reply develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md:181]: The idea behind it was that if system developers see that this is an option, know that they won't be able to use this just yet in any connector but that there's a minimum required version of 10.2.9.

File: user-guide/Reference/DataMiner_Extension_Modules.md (hunk 12 above)
Comment (mmphs) [reply user-guide/Reference/DataMiner_Extension_Modules.md]: Removed it as most of it was mentioned yet in DSI_OpenConfig_Middleware and extended that page with the things that were not mentioned yet.

File: develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md (hunk 11 above)
Comment (mmphs) [reply develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md]: Changed to 'it'.

File: develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md (hunk 8 above)
Comment (mmphs) [reply develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md:31]: Done

File: assemblies/DataSources/OpenConfig/Skyline.DataMiner.DataSources.OpenConfig.Gnmi.Protocol.xml [hunk 13]
```diff
@@ -0,0 +1,843 @@
...
+            <param name="pathElems">A collection of all path elements.</param>
+            <param name="updateValue">Value that needs to be changed.</param>
+            <param name="timestampUtc">UTC time when the value was taken.</param>
+            <param name="origin">Indication of where the data originated from.</param>
+            <param name="subscriptionName">Name of the subscription in case the data came in through subscription.</param>
+            <returns>Boolean indicating if there is something that needs to be changed.</returns>
+        </member>
+        <member name="M:Skyline.DataMiner.DataSources.OpenConfig.Gnmi.Protocol.DataMapper.DataMinerConnectorDataGrid.DeleteRow(System.String,System.DateTime)">
+            <summary>
+            Deletes the row when the timestamp is most recent.
```
Comment (MariekeGO) [inline assemblies/DataSources/OpenConfig/Skyline.DataMiner.DataSources.OpenConfig.Gnmi.Protocol.xml:173]: This is a slightly confusing sentence. Do you mean this?

```suggestion
            Deletes the row with the most recent timestamp.
```

Comment (mmphs) [reply assemblies/DataSources/OpenConfig/Skyline.DataMiner.DataSources.OpenConfig.Gnmi.Protocol.xml:173]:
```suggestion
            Deletes the row if <paramref name="timestampUtc"/> is more recent than the last received one.
```
It will delete the row if the timestamp that is passed along is more recent than the one that is stored in memory. If not it will actually not do anything.
